app/models/status.rb

```diff
@@ -188,12 +191,24 @@ def as_home_timeline(account)
       where(account: [account] + account.following).where(visibility: [:public, :unlisted, :private])
     end
 
-    def as_direct_timeline(account)
-      query = joins("LEFT OUTER JOIN mentions ON statuses.id = mentions.status_id AND mentions.account_id = #{account.id}")
-              .where("mentions.account_id = #{account.id} OR statuses.account_id = #{account.id}")
-              .where(visibility: [:direct])
+    def as_direct_timeline(account, max_id = nil, since_id = nil, limit = 20)
+      dm_to_me = Status.direct_messages.message_to_me(account).select(:id).limit(limit)
+      dm_from_me = Status.direct_messages.message_from_me(account).select(:id).limit(limit)
 
-      apply_timeline_filters(query, account, false)
+      if max_id.present?
+        dm_to_me = dm_to_me.where(arel_table[:id].lt(max_id))
+        dm_from_me = dm_from_me.where(arel_table[:id].lt(max_id))
+      end
+
+      if since_id.present?
+        dm_to_me = dm_to_me.where(arel_table[:id].gt(since_id))
+        dm_from_me = dm_from_me.where(arel_table[:id].gt(since_id))
+      end
+
+      dm_to_me = apply_timeline_filters(dm_to_me, account, false)
+      dm_from_me = apply_timeline_filters(dm_from_me, account, false)
+
+      Status.where(id: (dm_to_me + dm_from_me).uniq)
```

Comment:
dm_to_me and dm_from_me each gets filterd statuses with limit
-> There are 2 * limit filtered statuses
-> Line 211 gets ORDER BY id DESC LIMIT limit from 2 * limit filtered statuses
-> as_direct_timeline returns filterd statuses with limit correctly

```
mastodon_production=> explain analyze SELECT  "statuses"."id" FROM "statuses" LEFT OUTER JOIN "mentions" ON "mentions"."status_id" = "statuses"."id" LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."visibility" = 3 AND "mentions"."account_id" = 1 AND 1=1 AND "accounts"."silenced" = false ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                                         QUERY PLAN                                                                                         
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1.15..17869.62 rows=15 width=8) (actual time=0.080..0.523 rows=20 loops=1)
   ->  Nested Loop  (cost=1.15..17869.62 rows=15 width=8) (actual time=0.078..0.509 rows=20 loops=1)
         ->  Nested Loop  (cost=0.86..17787.92 rows=20 width=16) (actual time=0.071..0.427 rows=20 loops=1)
               ->  Index Only Scan Backward using index_mentions_on_account_id_and_status_id on mentions  (cost=0.42..2988.87 rows=1770 width=8) (actual time=0.021..0.105 rows=60 loops=1)
                     Index Cond: (account_id = 1)
                     Heap Fetches: 60
               ->  Index Scan using statuses_pkey on statuses  (cost=0.43..8.35 rows=1 width=16) (actual time=0.005..0.005 rows=0 loops=60)
                     Index Cond: (id = mentions.status_id)
                     Filter: (visibility = 3)
                     Rows Removed by Filter: 1
         ->  Index Scan using index_accounts_on_id on accounts  (cost=0.29..4.07 rows=1 width=8) (actual time=0.003..0.003 rows=1 loops=20)
               Index Cond: (id = statuses.account_id)
               Filter: (NOT silenced)
 Planning time: 0.508 ms
 Execution time: 0.559 ms
(15 rows)

mastodon_production=> explain analyze SELECT  "statuses"."id" FROM "statuses" LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."visibility" = 3 AND "statuses"."account_id" = 1 AND 1=1 AND "accounts"."silenced" = false ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                   QUERY PLAN                                                                    
-------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.85..88.47 rows=17 width=8) (actual time=0.043..1.021 rows=20 loops=1)
   ->  Nested Loop  (cost=0.85..88.47 rows=17 width=8) (actual time=0.042..1.011 rows=20 loops=1)
         ->  Index Only Scan using index_20180106 on statuses  (cost=0.56..79.95 rows=17 width=16) (actual time=0.031..0.976 rows=20 loops=1)
               Index Cond: ((account_id = 1) AND (visibility = 3))
               Heap Fetches: 20
         ->  Materialize  (cost=0.29..8.31 rows=1 width=8) (actual time=0.001..0.001 rows=1 loops=20)
               ->  Index Scan using index_accounts_on_id on accounts  (cost=0.29..8.31 rows=1 width=8) (actual time=0.008..0.009 rows=1 loops=1)
                     Index Cond: (id = 1)
                     Filter: (NOT silenced)
 Planning time: 0.170 ms
 Execution time: 1.054 ms
(11 rows)

imastodon=> explain analyze SELECT  "statuses"."id", "statuses"."updated_at" FROM "statuses" WHERE "statuses"."id"IN (xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx) ORDER BY "statuses"."id" DESC LIMIT 20;
                                                                                                                                                                                                                                                                                                                                                                                                         QUERY PLAN                                                                                                                                                                                                                                                                                                                                                                                                          
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.43..169.20 rows=20 width=16) (actual time=0.105..0.105 rows=0 loops=1)
   ->  Index Scan Backward using statuses_pkey on statuses  (cost=0.43..337.97 rows=40 width=16) (actual time=0.104..0.104 rows=0 loops=1)
         Index Cond: (id = ANY ('{xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx,xxxxxxxxxxxxxx}'::bigint[]))
 Planning time: 0.122 ms
 Execution time: 0.128 ms
(5 rows)
```